### PR TITLE
Faster regexp prefixes

### PR DESCRIPTION
**Pull request by bboreham**
Do a better job of regexps with a long series of choices followed by `.*`, like:
```
		"(?i:(zQPbMkNO.*|NNSPdvMi.*|iWuuSoAl.*|qbvKMimS.*|IecrXtPa.*|seTckYqt.*|NxnyHkgB.*|fIDlOgKb.*|UhlWIygH.*|OtNoJxHG.*|cUTkFVIV.*|mTgFIHjr.*|jQkoIDtE.*|PPMKxRXl.*|AwMfwVkQ.*|CQyMrTQJ.*|BzrqxVSi.*|nTpcWuhF.*|PertdywG.*|ZZDgCtXN.*|WWdDPyyE.*|uVtNQsKk.*|BdeCHvPZ.*|wshRnFlH.*|aOUIitIp.*|RxZeCdXT.*|CFZMslCj.*|AVBZRDxl.*|IzIGCnhw.*|ythYuWiz.*|oztXVXhl.*|VbLkwqQx.*|qvaUgyVC.*|VawUjPWC.*|ecloYJuj.*|boCLTdSU.*|uPrKeAZx.*|hrMWLWBq.*|JOnUNHRM.*|rYnujkPq.*|dDEdZhIj.*|DRrfvugG.*|yEGfDxVV.*|YMYdJWuP.*|PHUQZNWM.*|AmKNrLis.*|zTxndVfn.*|FPsHoJnc.*|EIulZTua.*|KlAPhdzg.*|ScHJJCLt.*|NtTfMzME.*|eMCwuFdo.*|SEpJVJbR.*|cdhXZeCx.*|sAVtBwRh.*|kVFEVcMI.*|jzJrxraA.*|tGLHTell.*|NNWoeSaw.*|DcOKSetX.*|UXZAJyka.*|THpMphDP.*|rizheevl.*|kDCBRidd.*|pCZZRqyu.*|pSygkitl.*|SwZGkAaW.*|wILOrfNX.*|QkwVOerj.*|kHOMxPDr.*|EwOVycJv.*|AJvtzQFS.*|yEOjKYYB.*|LizIINLL.*|JBRSsfcG.*|YPiUqqNl.*|IsdEbvee.*|MjEpGcBm.*|OxXZVgEQ.*|xClXGuxa.*|UzRCGFEb.*|buJbvfvA.*|IPZQxRet.*|oFYShsMc.*|oBHffuHO.*|bzzKrcBR.*|KAjzrGCl.*|IPUsAVls.*|OGMUMbIU.*|gyDccHuR.*|bjlalnDd.*|ZLWjeMna.*|fdsuIlxQ.*|dVXtiomV.*|XxedTjNg.*|XWMHlNoA.*|nnyqArQX.*|opfkWGhb.*|wYtnhdYb.*))",
```

### CODE REVIEW

File: model/labels/regexp.go [hunk 1]
```diff
@@ -206,38 +213,48 @@ func findSetMatchesInternal(re *syntax.Regexp, base string) (matches []string, c
...
 
 	matches = []string{base}
 
 	for i := 0; i < len(re.Sub); i++ {
-		var newMatches []string
+		// An OpStar makes the matches so far into prefixes.
+		if i > 0 && re.Sub[i].Op == syntax.OpStar && len(matches) > 0 {
+			prefixes = make([]prefixMatcher, 0, len(matches))
+			for _, prefix := range matches {
+				right := stringMatcherFromRegexpInternal(re.Sub[i])
```
[the remark this answers is not in the document]
Comment (pracucci) [reply model/labels/regexp.go:244]: For example this test fails:

```
	const pattern = "a.*a|b.*b|c.*c|d.*d|e.*e|f.*f|g.*g|h.*h|i.*i|l.*l|m.*m|n.*n|o.*o|p.*p|q.*q|r.*r|s.*s|t.*t|u.*u|v.*v|z.*z"

	standard := regexp.MustCompile(pattern)
	require.True(t, standard.MatchString("aa"))
	require.True(t, standard.MatchString("aXa"))
	require.False(t, standard.MatchString("aXb"))

	matcher, err := newFastRegexMatcherWithoutCache(pattern)
	require.NoError(t, err)
	require.True(t, matcher.MatchString("aa"))
	require.True(t, matcher.MatchString("aXa"))
	require.False(t, matcher.MatchString("aXb"))
```

**Review by pracucci — COMMENTED** [review]
This is an interesting PR. I think the overall idea is nice, but I think I've found a potential issue. Generally speaking, I think we're lacking some tests around edge cases that could be introduced by these changes. Also the fact that I've run fuzzy tests for 5m and they passed, but my simple test has failed is a bit concerning about the tests coverage we have here :|
